README.md
# Scribble Border

Scribble Border creates a visual border effect around a provided Widget. The aim of this border is to look like a user has drawn a rectangle or round shape around the provided child.
Since this is not a real Border it can be used around every visual Widget. Scribble Border also has a BackgroundPaint class, which can be used to fill a background with a Gradient.

|![Example1](https://github.com/NicolasDurant/scribble_border/blob/master/example/assets/Simulator%20Screen%20Shot%20-%20iPhone%2012%20-%202022-01-25%20at%2016.57.20.png)|![Example2](https://github.com/NicolasDurant/scribble_border/blob/master/example/assets/Simulator%20Screen%20Shot%20-%20iPhone%2012%20-%202022-01-25%20at%2016.57.13.png) | ![Example3](https://github.com/NicolasDurant/scribble_border/blob/master/example/assets/Simulator%20Screen%20Shot%20-%20iPhone%2012%20-%202022-01-25%20at%2017.06.15.png)|
|:---:|:---:|:---:|
## Getting Started

Add [scribble_border](https://pub.dev/packages/scribble_border) as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/).
You can also check the [installing page](https://pub.dev/packages/scribble_border/install).
Check out the [example page](https://pub.dev/packages/scribble_border/example) for a more detailed example or down below, for a simple example implementation of the package.

## Usage

Import the library via
```dart
import 'package:scribble_border/scribble_border.dart';
```

### Example
Refer to the [example.md](https://github.com/NicolasDurant/scribble_border/blob/master/example/example.md) to see how the screenshots are implemented.
```dart
import 'package:flutter/material.dart';
import 'package:scribble_border/scribble_border.dart';

class Example extends StatelessWidget {

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: ScribbleBorder(
          child: Container(
            height: 100,
            width: 100,
            color: Colors.white,
          ),
        ),
      ),
    );
  }
}
```

## ScribbleBorder - Supported Parameters
| named parameter             | type           | required | default                                           | description                                                                                                                                                                                                                                                                                        |
|-----------------------------|----------------|----------|---------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| child                       | Widget         | yes      |                                                   | Child to be surrounded by the [ScribbleBorder]                                                                                                                                                                                                                                                            |
| circular  | bool            | no       | false                                                 | If this is set the border will be drawn in a circle shape around the component                                                                                                                                                                                                                                |
| circularRadius     | double         | no       | 0.0                                               | Sets a custom radius for the circle. (only does anything, if circular = true). The default will create a circle with half the width of the component as radius                                                                                                                                                                                                                                   |
| gradient                 | Gradient | no      |         LinearGradient(colors: [Colors.black26, Colors.black12]                                          | The gradient is used for the color-range of a stroke and is always linear for now.                                                                                                                                                                             |
| gradientStretch                 | double         | no       | 0.5                                           | Value that determines how far the gradient of a stroke is stretched out. Bigger values mean less strong gradient. Always from the first color provided in the gradient to the second one. 0.0 means default linear gradient with even distribution.                                                                                                                                                                                                                                                                 |
| padding            | double         | no       | 0                                               | Padding for the [Container] that surrounds the [child]                                                                                                                                                                              |
| painter             | ScribblePainter         | no       |                                                |  Our custom paint to be applied on the border. You don't need to set this                                                                                                                                                                                             |
| randomOffset        | double          | no       | 0.0                                      | Value that determines how far strokes can overlap or how big the gaps can be. Bigger values mean bigger gaps and overlaps. 0.0 means no offset.     |                                                                                                                                                                                                                                                        |
| repaint             | bool         | no       | false                                               | If this is set, the component will repaint when the build method of the parent class is called. Since [ScribbleBorder] works with random values this should be false for the majority of the use cases                                                                                                                                                                                                          |
| strokeWidth | double          | no       | 3.0                                  | Size of the drawn lines                                                                                                                                                                                                                                    |

## BackgroundPaint - Supported Parameters
| named parameter             | type           | required | default                                           | description                                                                                                                                                                                                                                                                                        |
|-----------------------------|----------------|----------|---------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| child                       | Widget         | yes      |                                                   | Child to be surrounded by the [BackgroundPaint]                                                                                                                                                                                                                                                            |
| corners                       | Corners         | no      |            Corners.all(25)                                       | Consisting of 4x [Radius] to determine the radius of the 4 corners of the rectangle to draw. Setting all to 100 will create a circular shape                                                                                                                                                                                                                                                            |
| gradient                 | Gradient | no      |         RadialGradient(colors: [Color.fromARGB(200, 232, 53, 44), Color.fromARGB(255, 224, 15, 4),],)                                          | The gradient is used for the filling color of the background. We take a RadialGradient, which works great with rectangle filling                                                                                                                                                                             |
| padding            | double         | no       | 0                                               | Padding for the [Container] that surrounds the [child]                                                                                                                                                                              |
| painter             | BackgroundPainter         | no       |                                                |  Our custom paint to be applied on the border. You don't need to set this                                                                                                                                                                                             |
| randomOffset        | double          | no       | 0.0                                      | Value that determines how far strokes can overlap or how big the gaps can be. Bigger values mean bigger gaps and overlaps. 0.0 means no offset.     |                                                                                                                                                                                                                                                        |
| randomRadius        | int          | no       | 0.0                                      | Value that determines how far corner radius can randomize. Bigger values mean bigger differences. 0 means no random corners.     |                                                                                                                                                                                                                                                        |
| repaint             | bool         | no       | false                                               | If this is set, the component will repaint when the build method of the parent class is called. Since [ScribbleBorder] works with random values this should be false for the majority of the use cases                                                                                                                                                                                                          |

CHANGELOG.md
## 1.0.0

* Initial release.
* Add ScribbleBorder
* Add BackgroundPaint
* Add documentation
* Add example project


test/scribble_border_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:scribble_border/scribble_border.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/src/scribble_border.dart
import 'package:flutter/material.dart';
import 'package:scribble_border/src/scribble_painter.dart';

/// [ScribbleBorder] uses the [ScribblePainter] classes [CustomPainter] to create a visual border effect around a provided [Widget].
/// The aim of this border is to look like a user has drawn a rectangle shape around the provided [child].
/// Since this is not a real [Border] it can be used around every visual [Widget].
class ScribbleBorder extends StatelessWidget {
  ScribbleBorder(
      {Key? key,
      required this.child,
      this.gradient = const LinearGradient(colors: [Colors.black26, Colors.black12]),
      this.padding = 0,
      this.randomOffset = 0.0,
      this.gradientStretch = 0.5,
      this.strokeWidth = 3,
      this.circular = false,
      this.repaint = false,
      this.circularRadius = 0.0})
      : painter = ScribblePainter(
          gradient: gradient,
          strokeWidth: strokeWidth,
          randomOffset: randomOffset,
          gradientStretch: gradientStretch,
          circular: circular,
          circularRadius: circularRadius,
          repaint: repaint,
        ),
        super(key: key);

  /// Child to be surrounded by the [ScribbleBorder].
  final Widget child;

  /// The gradient is used for the color-range of a stroke and is always linear for now.
  final Gradient gradient;

  /// Padding for the [Container] that surrounds the [child]..
  final double padding;

  /// Value that determines how far strokes can overlap or how big the gaps can be.
  /// Bigger values mean bigger gaps and overlaps.
  /// 0.0 means no offset.
  final double randomOffset;

  /// Value that determines how far the gradient of a stroke is stretched out.
  /// Bigger values mean less strong gradient. Always from the first color provided in the gradient to the second one.
  /// 0.0 means default linear gradient with even distribution.
  final double gradientStretch;

  /// Size of the drawn lines.
  final double strokeWidth;

  /// If this is set the border will be drawn in a circle shape around the component.
  final bool circular;

  /// Sets a custom radius for the circle. (only does anything, if circular = true).
  /// The default will create a circle with half the width of the component as radius.
  final double circularRadius;

  /// If this is set, the component will repaint when the build method of the parent class is called.
  /// Since [ScribbleBorder] works with random values this should be false for the majority of the use cases.
  final bool repaint;

  /// Our custom paint to be applied on the border.
  final ScribblePainter painter;

  @override
  Widget build(BuildContext context) {
    return CustomPaint(
      painter: painter,
      child: Container(
        padding: EdgeInsets.all(padding),
        child: child,
      ),
    );
  }
}


lib/src/scribble_painter.dart
import 'dart:math';
import 'dart:ui' as ui;

import 'package:flutter/material.dart';

/// [ScribblePainter] serves as a [CustomPainter] for the [ScribbleBorder] class.
/// It will draw lines in a rectangle shape around the provided [Widget] in [ScribbleBorder].
/// It takes the size of a stroke [strokeWidth] and a [gradient] to determine its color.
/// [ScribbleBorder] uses random values to let the strokes appear to be drawn by a human with imperfect edges.
class ScribblePainter extends CustomPainter {
  ScribblePainter({
    required this.gradient,
    required this.randomOffset,
    required this.gradientStretch,
    required this.strokeWidth,
    required this.circular,
    required this.circularRadius,
    required this.repaint,
  });

  /// The gradient that should be used upon the strokes. This will always be linear for now and only takes the colors.
  final Gradient gradient;

  /// Value that determines how far strokes can overlap or how big the gaps can be.
  /// Bigger values mean bigger gaps and overlaps.
  /// 0.0 means no offset.
  final double randomOffset;

  /// Value that determines how far the gradient of a stroke is stretched out.
  /// Bigger values mean less strong gradient.
  /// 0.0 means default linear gradient with even distribution.
  final double gradientStretch;

  /// Size of the drawn lines.
  final double strokeWidth;

  /// If this is set the border will be drawn in a circle shape around the component.
  final bool circular;

  /// Sets a custom radius for the circle. (only does anything, if circular = true).
  /// The default will create a circle with half the width of the component as radius.
  final double circularRadius;

  /// If this is set, the component will repaint when the build method of the parent class is called.
  /// Since [ScribbleBorder] works with random values this should be false for the majority of the use cases.
  final bool repaint;

  @override
  void paint(Canvas canvas, Size size) {
    // create the visual appearance of the drawn lines
    Paint paintObject = Paint();
    paintObject.style = PaintingStyle.stroke;
    paintObject.strokeWidth = strokeWidth;
    // create a bigger range to have double values
    int offsetMin = -(randomOffset * 100).toInt();
    int offsetMax = (randomOffset * 100).toInt();
    int strokeMin = -(gradientStretch * 100).toInt();
    int strokeMax = (gradientStretch * 100).toInt();
    if (circular) {
      // points for gradient
      Offset p1 = Offset.zero;
      Offset p2 = Offset(size.width, size.height);
      // middle point
      Offset middle = Offset(size.width / 2 - randomNumber(offsetMin, offsetMax, true) / 100,
          size.height / 2 - randomNumber(offsetMin, offsetMax, true) / 100);
      // gradient shader
      paintObject.shader = ui.Gradient.linear(
          Offset(p1.dx * randomNumber(strokeMin, strokeMax, false) / 100,
              p1.dy * randomNumber(strokeMin, strokeMax, false) / 100),
          Offset(p2.dx * randomNumber(strokeMin, strokeMax, false) / 100,
              p2.dy * randomNumber(strokeMin, strokeMax, false) / 100),
          gradient.colors);
      // draw circle
      canvas.drawCircle(middle, circularRadius != 0.0 ? circularRadius : size.height / 2, paintObject);
    } else {
      // points of the rectangle
      Offset p1 = Offset.zero;
      Offset p2 = Offset(size.width, 0);
      Offset p3 = Offset(0, size.height);
      Offset p4 = Offset(size.width, size.height);
      // draw p1 -> p2
      paintObject.shader = ui.Gradient.linear(
          p1,
          Offset(p2.dx * randomNumber(strokeMin, strokeMax, false) / 100,
              p2.dy * randomNumber(strokeMin, strokeMax, false) / 100),
          gradient.colors);
      canvas.drawLine(
          Offset(p1.dx - randomNumber(offsetMin, offsetMax, true) / 100,
              p1.dy - randomNumber(offsetMin, offsetMax, true) / 100),
          Offset(p2.dx - randomNumber(offsetMin, offsetMax, true) / 100,
              p2.dy - randomNumber(offsetMin, offsetMax, true) / 100),
          paintObject);
      // draw p1 -> p3
      paintObject.shader = ui.Gradient.linear(
          p1,
          Offset(p3.dx * randomNumber(strokeMin, strokeMax, false) / 100,
              p3.dy * randomNumber(strokeMin, strokeMax, false) / 100),
          gradient.colors);
      canvas.drawLine(
          Offset(p1.dx - randomNumber(offsetMin, offsetMax, true) / 100,
              p1.dy - randomNumber(offsetMin, offsetMax, true) / 100),
          Offset(p3.dx - randomNumber(offsetMin, offsetMax, true) / 100,
              p3.dy - randomNumber(offsetMin, offsetMax, true) / 100),
          paintObject);
      // draw p2 -> p4
      paintObject.shader = ui.Gradient.linear(
          p2,
          Offset(p4.dx * randomNumber(strokeMin, strokeMax, false) / 100,
              p4.dy * randomNumber(strokeMin, strokeMax, false) / 100),
          gradient.colors);
      canvas.drawLine(
          Offset(p2.dx - randomNumber(offsetMin, offsetMax, true) / 100,
              p2.dy - randomNumber(offsetMin, offsetMax, true) / 100),
          Offset(p4.dx - randomNumber(offsetMin, offsetMax, true) / 100,
              p4.dy - randomNumber(offsetMin, offsetMax, true) / 100),
          paintObject);
      // draw p3 -> p4
      paintObject.shader = ui.Gradient.linear(
          p3,
          Offset(p4.dx * randomNumber(strokeMin, strokeMax, false) / 100,
              p4.dy * randomNumber(strokeMin, strokeMax, false) / 100),
          gradient.colors);
      canvas.drawLine(
          Offset(p3.dx - randomNumber(offsetMin, offsetMax, true) / 100,
              p3.dy - randomNumber(offsetMin, offsetMax, true) / 100),
          Offset(p4.dx - randomNumber(offsetMin, offsetMax, true) / 100,
              p4.dy - randomNumber(offsetMin, offsetMax, true) / 100),
          paintObject);
    }
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) => repaint;

  /// Takes a [min] and [max] value as range for the random generated number.
  /// This only works with [int] values, which is why the random numbers for the borders in this class
  /// get multiplied before being applied to the function, and divided afterwards.
  double randomNumber(int min, int max, bool offset) {
    if (min == 0 && max == 0) return offset ? 0.0 : 100;
    var random = new Random();
    return (min + random.nextInt(max - min)).toDouble();
  }
}


lib/src/background_painter.dart
import 'dart:math';

import 'package:flutter/material.dart';

import '../scribble_border.dart';

/// [BackgroundPainter] serves as a [CustomPainter] for the [BackgroundPaint] class.
/// It will draw lines in a rectangle shape around the provided [Widget] in [BackgroundPaint].
/// It takes the size of a [Rect] around provided [child] and a [gradient] to determine its color. [RadientGradient] should be chosen.
/// [BackgroundPaint] uses random values to let the rectangle appear to be drawn by a human with imperfect edges and [Radius] x [Corners].
class BackgroundPainter extends CustomPainter {
  BackgroundPainter({
    required this.corners,
    required this.gradient,
    required this.randomOffset,
    required this.randomRadius,
    required this.repaint,
  });

  // Consisting of 4x [Radius] to determine the radius of the 4 corners of the rectangle to draw.
  final Corners corners;

  /// The gradient is used for the filling color of the background.
  /// We take a RadialGradient, which works great with rectangle filling.
  final Gradient gradient;

  /// Value that determines how far the rectangle can overlap or how big the gaps can be.
  /// Bigger values mean bigger gaps and overlaps of the rectangle..
  /// 0 means no offset.
  final int randomOffset;

  /// Value that determines how far corner radius can randomize.
  /// Bigger values mean bigger differences.
  /// 0 means no random corners.
  final int randomRadius;

  /// If this is set, the component will repaint when the build method of the parent class is called.
  /// Since [BackgroundPaint] works with random values this should be false for the majority of the use cases.
  final bool repaint;

  @override
  void paint(Canvas canvas, Size size) {
    Paint paintObject = Paint();
    // create a bigger range to have double values
    int min = randomOffset * -100;
    int max = randomOffset * 100;
    int radiusMin = randomRadius * -100;
    int radiusMax = randomRadius * 100;
    // create the rounded rectangle to be drawn
    RRect innerRect = RRect.fromLTRBAndCorners(
      0 - randomNumber(min, max) / 100,
      0 - randomNumber(min, max) / 100,
      size.width - randomNumber(min, max) / 100,
      size.height - randomNumber(min, max) / 100,
      topLeft: Radius.elliptical(corners.topLeft.x - randomNumber(radiusMin, radiusMax) / 100,
          corners.topLeft.y - randomNumber(radiusMin, radiusMax) / 100),
      topRight: Radius.elliptical(corners.topRight.x - randomNumber(radiusMin, radiusMax) / 100,
          corners.topRight.y - randomNumber(radiusMin, radiusMax) / 100),
      bottomLeft: Radius.elliptical(corners.bottomLeft.x - randomNumber(radiusMin, radiusMax) / 100,
          corners.bottomLeft.y - randomNumber(radiusMin, radiusMax) / 100),
      bottomRight: Radius.elliptical(corners.bottomRight.x - randomNumber(radiusMin, radiusMax) / 100,
          corners.bottomRight.y - randomNumber(radiusMin, radiusMax) / 100),
    );
    // create the gradient shade filler
    paintObject.shader =
        gradient.createShader(Rect.fromLTRB(-size.width, -size.height, size.width * 2, size.height * 2));
    canvas.drawRRect(innerRect, paintObject);
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) => repaint;

  /// Takes a [min] and [max] value as range for the random generated number.
  /// This only works with [int] values, which is why the random numbers for the borders in this class
  /// get multiplied before being applied to the function, and divided afterwards.
  double randomNumber(int min, int max) {
    if (min == 0 && max == 0) return 0.0;
    var random = new Random();
    return (min + random.nextInt(max - min)).toDouble();
  }
}


lib/src/background_paint.dart
import 'package:flutter/material.dart';

import 'background_painter.dart';

/// [BackgroundPaint] uses the [BackgroundPainter] classes [CustomPainter] to create a visual rectangle background effect underneath a provided [Widget].
/// The aim of this background is to look like a user has drawn a rectangle shape under the provided [child].
/// Since this is not a real background it can be used around every visual [Widget].
class BackgroundPaint extends StatelessWidget {
  BackgroundPaint({
    Key? key,
    required this.child,
    this.gradient = const RadialGradient(
      colors: [
        Color.fromARGB(200, 232, 53, 44),
        Color.fromARGB(255, 224, 15, 4),
      ],
    ),
    this.randomOffset = 0,
    this.randomRadius = 0,
    this.padding = 0,
    this.corners = const Corners(
      Radius.circular(25),
      Radius.circular(25),
      Radius.circular(25),
      Radius.circular(25),
    ),
    this.repaint = false,
  })  : painter = BackgroundPainter(
          gradient: gradient,
          corners: corners,
          randomOffset: randomOffset,
          randomRadius: randomRadius,
          repaint: repaint,
        ),
        super(key: key);

  /// Child to be surrounded by the [BackgroundPaint].
  final Widget child;

  /// The gradient is used for the filling color of the background.
  /// We take a RadialGradient, which works great with rectangle filling.
  final Gradient gradient;

  // Consisting of 4x [Radius] to determine the radius of the 4 corners of the rectangle to draw. Setting all to 100 will create a circular shape.
  final Corners corners;

  /// Padding for the [Container] that surrounds the [child].
  final double padding;

  /// Value that determines how far strokes can overlap or how big the gaps can be.
  /// Bigger values mean bigger gaps and overlaps.
  /// 0 means no offset.
  final int randomOffset;

  /// Value that determines how far corner radius can randomize.
  /// Bigger values mean bigger differences.
  /// 0 means no random corners.
  final int randomRadius;

  /// If this is set, the component will repaint when the build method of the parent class is called.
  /// Since [BackgroundPaint] works with random values this should be false for the majority of the use cases.
  final bool repaint;

  /// Our custom paint to be applied on the border.
  final BackgroundPainter painter;

  @override
  Widget build(BuildContext context) {
    return CustomPaint(
      painter: painter,
      child: Container(
        padding: EdgeInsets.all(padding),
        child: child,
      ),
    );
  }
}

class Corners {
  final Radius topLeft;
  final Radius topRight;
  final Radius bottomLeft;
  final Radius bottomRight;

  const Corners(this.topLeft, this.topRight, this.bottomLeft, this.bottomRight);

  Corners.all(double radius)
      : this.topLeft = Radius.circular(radius),
        this.topRight = Radius.circular(radius),
        this.bottomLeft = Radius.circular(radius),
        this.bottomRight = Radius.circular(radius);
}
